platform/engineering/design-docs/2021-05-18-forms-library.md

```diff
@@ -0,0 +1,852 @@
+# Forms library
+
+**Author(s):** Christopher Valarida  
+**Last Updated:** May 18, 2021  
+**Status:** **Draft** | In Review | Approved  
+**Approvers:** Brooks Johnson [ ], Michael Fleet [ ]  
+
+## Overview
+
+### Objective
+Create a simple, flexible forms library which can be used either in whole or in
+part by application teams building any kind of form on VA.gov.
+
+The intended audience of this document is
+- Front end engineers building the library
+- Application teams who may use it to build forms on VA.gov
+
+### Background
+We currently have a library which application teams use to build form
+applications on VA.gov. It employs a configuration-based approach to rapidly
+build forms in a consistent way.
+
+To distinquish the new forms library from the old in this document, this new
+forms library will be called **Formulate**.
+
+#### Problems
+
+##### Library bloat
+The current forms library uses a single form configuration object to create the
+entire form application. Modifying this `formConfig` object is the only way to
+change anything in the form. When new features for individual forms are needed,
+a change to the underlying library is required. This creates significant bloat
+in the library and makes it difficult to modify.
+
+##### Inflexibility
+Often, an application team will want to tweak an aspect of a form, but the only
+way to do that is by modifying the underlying library. See also [Library
+bloat](#library-bloat) above.
+
+##### Steep learning curve
+The book _Good Services_ sets forth a couple of relevant service design
+principles:
+- Work in a way that is familiar
+- Require no prior knowledge to use
+
+The form configuration approach is not one that is familiar to front end
+engineers to create applications and requires _much_ prior knowledge to use
+effectively. When learning the basics of the library, its complexity is
+relatively minimal, but quickly grows as the form becomes more complex. This is
+largely due to the library bloat outlined above.
+
+##### And more!
+There are many other problems introduced by the existing forms library, but for
+now, I digress.
```

Comment:
I would list them if possible, that way you'll have a more concrete problem statement

Reply:
Yeah, that's a good call. Only trouble is it's been so long since I've used the library, I can't remember them all, so this is where I phone a friend or...several. :sweat_smile: 

What other problems do you run into, @jbalboni, @Mottie, @LinnJS (and whomever else)?

I know there are some underlying assumptions that aren't always true for all forms these days:
- Must have an introduction page
- Must have a review page
- Must have a schema
- ...what else?

Reply:
- The path must always start on the introduction page (wizards may need their own path eventually)
- Schema in vets-json-schema isn't the same schema you'll use to build the form
- Form validation errors based on schema do not match visual representation
- List loop (ArrayField) component isn't customizable - I've copy/pasted and modified the same component into multiple forms now
- Checkbox error messaging not working as expected
- Checkbox group pattern does not handle error messages

And these are just the ones I've encountered recently...

Reply:
I added a bunch of bullet points based on your feedback. Thanks kindly! If there are any more you'd like to call out, please leave a comment in your review. :grin: